Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m

```diff
@@ -0,0 +1,356 @@
+        sideCarMat = fullfile(Fpath, [Ffile '.mat']);
+    
+        if exist(sideCarMat, 'file')
+            % we check if the mat-sidecar exists, with the motion estimation parameters
+            % If this file exists, we assume that the original ASL4D input file has not been
+            % interpolated yet (and the motion estimation has not been applied yet)
+    
+            %% ------------------------------------------------------------------------------------------
+            % Apple motion correction & resample
+            if nVolumes>1
+                for iS=1:nVolumes
+                    matlabbatch{1}.spm.spatial.realign.write.data{iS,1} = [path_ASL4D ',' num2str(iS)];
+                end
+            
+                matlabbatch{1}.spm.spatial.realign.write.roptions.which     = [2 0];
+                matlabbatch{1}.spm.spatial.realign.write.roptions.interp    = 4;
+                matlabbatch{1}.spm.spatial.realign.write.roptions.wrap      = [0 0 0];
+                matlabbatch{1}.spm.spatial.realign.write.roptions.mask      = 1;
+                matlabbatch{1}.spm.spatial.realign.write.roptions.prefix    = 'r';
+            
+                spm_jobman('run',matlabbatch); % this applies the motion correction in native space
+                
+                path_rASL4D = fullfile(Fpath, ['r' Ffile '.nii']);
+                % Reload motion corrected timeseries:
+                ASL4D = xASL_io_Nifti2Im(path_rASL4D);
+                xASL_delete(path_rASL4D);
+            end
+        end
+    end
+end
+    
+% =====================================================================
+% -> THIS WILL BE MOVED TO xASL_quant_HadamardDecoding
+%% 3. Time-Encoded subtraction
+% xASL_quant_HadamardDecoding should not only decode the image volumes but also the quantification parameters
+% that accompany the different volumes, i.e. EchoTime, PLD, and LabelingDuration.
+% When this is done appropriately, PWI derivative volumes are easily created based on these parameters
+
+if x.modules.asl.bTimeEncoded
+    
+    if bCreatePWI4D
+        %% 1) Create PWI4D
+        % Decoding of TimeEncoded data - it outputs a decoded image and it also saves as a NII
+        PWI4D = xASL_quant_HadamardDecoding(ASL4D, x.Q);
+	    nVolumes = size(PWI4D, 4);
+    end
+
+    if bCreatePWI3D
+
+    %% 2) Create PWI3D
+	% Calculate Hadamard block size (number of unique PLDs.*labeling durations.* echo times) = number of volumes per repetition
+    
+
+    %% 1) x.Q.EchoTime vectors (identical to below)
+    % First we deal with a too long vector (e.g., cutting off control volumes for Hadamard)
+    nTE = length(x.Q.EchoTime);
+    if nTE>nVolumes
+        x.Q.EchoTime_PWI4D = x.Q.EchoTime(1:nVolumes); % can we do this based on the ASL4Dcontext.tsv ?
+    else
+        x.Q.EchoTime_PWI4D = x.Q.EchoTime;
+    end
+    nTE = length(x.Q.EchoTime_PWI4D);
+
+    
+    nUniqueTE = length(uniqueTE); % Obtain the number of echo times
```

Comment:
These parameters are already created in the ASL module.
x.Q.uniqueEchoTimes = uniquetol(x.Q.EchoTime, 0.001);
    x.Q.NumberEchoTimes = length(x.Q.uniqueEchoTimes);